include/mrdox/Support/Handlebars.hpp

```diff
@@ -0,0 +1,1037 @@
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+
+#ifndef MRDOX_TOOL_SUPPORT_PATH_HPP
+#define MRDOX_TOOL_SUPPORT_PATH_HPP
+
+#include <mrdox/Support/String.hpp>
+#include <mrdox/Support/Dom.hpp>
+#include <string_view>
+#include <unordered_map>
+#include <functional>
+#include <type_traits>
+#include <vector>
+
+namespace clang {
+namespace mrdox {
+
+/** Options for handlebars
+
+    In particular, we have the noHTMLEscape option, which we
+    use to disable HTML escaping when rendering asciidoc templates.
+
+    This struct is analogous to the Handlebars.compile options.
+
+ */
+struct HandlebarsOptions
+{
+    /** Escape HTML entities
+     */
+    bool noHTMLEscape = false;
+
+    /** Templates will throw rather than ignore missing fields
+     */
+    bool strict = false;
+
+    /** Disable the auto-indent feature
+
+        By default, an indented partial-call causes the output of the
+        whole partial being indented by the same amount.
+     */
+    bool preventIndent = false;
+
+    /** Disables standalone tag removal when set to true
+
+        When set, blocks and partials that are on their own line will not
+        remove the whitespace on that line
+     */
+    bool ignoreStandalone = false;
+};
+
+// Objects such as llvm::raw_string_ostream
+template <typename Os>
+concept LHROStreamable =
+requires(Os &os, std::string_view sv)
+{
+    { os << sv } -> std::convertible_to<Os &>;
+};
+
+// Objects such as std::ofstream
+template <typename Os>
+concept StdLHROStreamable = LHROStreamable<Os> && std::convertible_to<Os*, std::ostream*>;
+
+// Objects such as std::string
+template <typename St>
+concept SVAppendable =
+requires(St &st, std::string_view sv)
+{
+    st.append( sv.data(), sv.data() + sv.size() );
+};
+
+/** Reference to output stream used by handlebars
+
+    This class is used to internally pass an output stream to the
+    handlebars engine.
+
+    It allows many types to be used as output streams, including
+    std::string, std::ostream, llvm::raw_string_ostream, and others.
+
+ */
+class MRDOX_DECL OutputRef
+{
+    friend class Handlebars;
+
+    using fptr = void (*)(void * out, std::string_view sv);
```

Comment:
This is one of the reasons I rejected Peter's Mustache submission...because it is push based rather than pull based. It can't handle the simple use-case of filling a buffer to capacity and then resuming later...

Reply:
I didn't give this much thought. The previous version was appending to a string and this version was just an incremental improvement because I needed to append to other things at some points. 

What are the alternatives?

Reply:
Would you believe that I was thinking of `co_await` ? :)

Reply:
Well... that's really another level of complexity. :)